lib/elixir/lib/enum.ex

```diff
@@ -3236,21 +3242,31 @@ defmodule Enum do
       iex> Enum.with_index([:a, :b, :c], 3)
       [a: 3, b: 4, c: 5]
 
+      iex> Enum.with_index([:a, :b, :c], fn element, index -> {index, element} end)
+      [{0, :a}, {1, :b}, {2, :c}]
   """
-  @spec with_index(t, integer) :: [{element, index}]
-  def with_index(enumerable, offset \\ 0) do
+  @spec with_index(t, integer | (element, index -> any)) :: list
+  def with_index(enumerable, fun_or_offset \\ 0)
+
+  def with_index(enumerable, offset) when is_integer(offset) do
+    enumerable
+    |> to_list()
+    |> do_with_index(offset, fn element, index -> {element, index} end)
+  end
+
+  def with_index(enumerable, fun) when is_function(fun, 2) do
     enumerable
     |> to_list()
-    |> do_with_index(offset)
+    |> do_with_index(0, fun)
   end
 
-  @spec do_with_index(list, integer) :: [{element, index}]
-  defp do_with_index([], _) do
+  @spec do_with_index(list, integer, function) :: list
+  defp do_with_index([], _offset, _fun) do
     []
   end
 
-  defp do_with_index([head | tail], index) do
-    [{head, index} | do_with_index(tail, index + 1)]
+  defp do_with_index([head | tail], index, fun) do
+    [fun.(head, index) | do_with_index(tail, index + 1, fun)]
   end
```

Comment:
The implementation we have here is not very efficient unfortunately. For things that are not lists, it traverses everything twice. I think we should write it like this:

```elixir
def with_index(enumerable, offset) when is_integer(offset) do
  enumerable
  |> Enumerable.map_reduce(offset, fn x, i -> {{x, i}, i+1} end)
  |> elem(0)
end

def with_index(enumerable, fun) when is_function(fun) do
  enumerable
  |> Enumerable.map_reduce(offset, fn x, i -> {fun.(x, i), i+1} end)
  |> elem(0)
end
```

Can you please update accordingly?

Thank you.

Reply:
Sure!

Reply:
Just pushed the updated version, @josevalim :)